tests/lang_callable/iter/titer_issues.nim

```diff
@@ -250,3 +250,31 @@ block:
 
   for x in ff(@[1, 2], @[1, 2, 3]):
     echo x
+
+block addr_of_parameters:
+  # iterator parameters support having their address taken, regardless of how
+  # the argument expression looks like
+  proc get(): int {.noinline.} = 2
+
+  iterator iter(a: int, b: int): (ptr int, ptr int) {.inline.} =
+    yield (addr(a), addr(b))
+
+  for a, b in iter(1, get()):
+    doAssert a[] == 1
+    doAssert b[] == 2
+
+block ref_construction_argument:
+  # a literal ref construction expression must only be evaluated once, prior
+  # to control-flow entering the iterators body
+  type RefObj = ref object
+    i: int
+
+  iterator iter(a: RefObj): int =
+    # use `a` multiple times in order to detect the argument expression
+    # being erroneously inlined
+    inc a.i
+    inc a.i
+    yield a.i
+
+  for i in iter(RefObj(i: 1)):
+    doAssert i == 3
```

Comment:
The problem here was that the argument passing logic treated an `nkObjConstr` node as something with no side-effects, as long as none of the field initializer expression had any. While this is true for the construction of normal objects, it is not true for when the result is of `ref` type. In effect, what happened here was that each occurrence of `a` in the iterator's body was replaced with the `nkObjConstr` AST, and the integer value yielded by the iterator was thus 1.

I've first thought about this while working on the MIR, but I think that given their different behaviour, it would make sense to have a dedicated node kind for `ref` object construction. Alternatively, the `nkObjConstr` nodes for them could be lowered into something like:
```nim
let x = new T
y[] = # object construction expression goes here
x
``` 
early on, but that might complicate some analysis a bit.

Reply:
Playing off what you're driving towards, I think `ref` construction is perhaps best thought of as:

- `ref`'s location allocation
- `ref`'d memory allocation + the consturctor expression as initialization
- assigning the `ref`'d memory's address to the `ref` location

In which case, the `nkObjectConstr` for a `ref` should be within some node that treats it as an initialization expression and not the usual "value construction". Or in other words, "value construction" fuses (combines & elides) the more granular steps that the `ref` variant takes, just a matter of perspective.